### PR TITLE
Add ContextSensitiveWeakHashMap

### PR DESCRIPTION
**Pull request by ruhan1**
This provides a ThreadContext sensitive WeakHashMap where an entry will not be GC-ed until it is cleared by all threads that use it.

### CODE REVIEW

**Comment by ruhan1** [comment]
retest this please

**Comment by ruhan1** [comment]
retest this please

**Comment by bxf12315** [comment]
retest this please

File: src/main/java/org/commonjava/cdi/util/weft/ThreadContext.java [hunk 1]
```diff
@@ -106,10 +106,10 @@ private static ThreadContext swapContext( final ThreadContext ctx )
      */
     public synchronized void registerFinalizer( Consumer<ThreadContext> finalizer )
     {
-        Logger logger = LoggerFactory.getLogger( getClass() );
-        logger.trace( "Registering finalizer: {} on ThreadContext: {}", finalizer, this );
         if ( !this.finalizers.contains( finalizer ) )
         {
+            Logger logger = LoggerFactory.getLogger( getClass() );
+            logger.debug( "Registering finalizer: {} on ThreadContext: {}", finalizer, this );
```
Comment (jdcasey) [inline src/main/java/org/commonjava/cdi/util/weft/ThreadContext.java]: We might want to put this back on TRACE log level, but not a huge problem for now.

File: src/main/java/org/commonjava/cdi/util/weft/ContextSensitiveWeakHashMap.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+/**
+ * ContextSensitiveWeakHashMap is a subclass of WeakHashMap. But an entry will not be GC-ed until all the threads
+ * ever use it clears their ThreadContext.
+ *
+ * An utility method newSynchronizedContextSensitiveWeakHashMap is provided as a convenient way to obtain a
+ * synchronized ContextSensitiveWeakHashMap instance.
+ *
+ * Created by ruhan on 12/21/17.
+ */
+public class ContextSensitiveWeakHashMap<K, V>
```
Comment (jdcasey) [inline src/main/java/org/commonjava/cdi/util/weft/ContextSensitiveWeakHashMap.java]: I'm having a little trouble following the logic of this. It seems like what we need is a variant of HashMap that can count the number of references for a given key, then when the ref count hits zero, removes the key. In Weft, I'd assume this would mean using a UUID per ThreadContext for use in the map (which has a lifecycle that's longer than any single ThreadContext)...then use that UUID to place a referent on each key when the key is added or accessed. Then, when the ThreadContext finalizer runs, it would remove that UUID referent from the key...then removes the key if that results in a zero reference count for that key. Again, this would be working on a map instance that exists **outside** the ThreadContext, but uses UUIDs which are associated with ThreadContexts to do reference counting.

It seems like this class does the inverse. It keeps a UUID for the whole map, and uses that UUID to keep a set of keys in the ThreadContext, removing them from the ThreadContext when the finalizer runs.

---

**OK** I'm leaving this comment here as a sign of respect. This is a really elegant solution...your GC-fu is strong. 😄 

Basically, you're using the Set in the ThreadContext as a way of making the keys strongly reachable, without really centrally managing the reference counts...the counts are implied by the set of contexts that keep strong references. Each one that runs its finalizers relinquishes its strong reference, and eventually this will allow GC to clear the WeakHashMap of those keys.

VERY nice.

Comment (ruhan1) [reply src/main/java/org/commonjava/cdi/util/weft/ContextSensitiveWeakHashMap.java]: You got it. Thanks for the review. 
It had been a hard time for me to come up with a good solution. I had thought about it for some days until I appreciated that ThreadContext is a great tool for implementing such "stronger weak-hashmap", and coded it within an hour.
